### DISCUSSION TITLE
Does ld link dynamically to the runtime libraries by default?

### DISCUSSION THREAD

**Discussion by TheShermanTanker** [msys2/MINGW-packages#11601] [Q&A]
Does the ucrt64 toolchain link dynamically to ucrtbase.dll and libstdc++.dll by default? Asking because the final binaries generated by the compiler are at the very least a whopping 133KB in size even for tiny programs, which leads me to believe that the libraries might be getting statically linked quietly (or if other Windows libraries such as user32 are getting pulled into the binary without me knowing), but I can't be sure since I don't have a way to check whether they are

**Comment by Biswa96** [comment]
> Does the ucrt64 toolchain by default link dynamically to ucrtbase.dll and libstdc++.dll by default?

Yes.

> the final binaries generated by the compiler are at the very least a whopping 133KB in size even for tiny programs

`strip` can make it smaller.

**Reply by TheShermanTanker** [reply]
Hmm, does the old trick of --gc-sections and -ffunction-sections/-fdata-sections work with the ucrt compiler in that case?

**Reply by Biswa96** [reply]
I am not deeply familiar with those options. But with my quick poking, it looks like `--gc-sections` works but `-ffunction-sections` or `-fdata-sections` does not. Other experts can provide some details. You can investigate with mapfile and `--print-gc-sections` options.

**Reply by TheShermanTanker** [reply]
> Yes.

Strangely when I try running the compiled binaries with libstdc++.dll in a directory where it can't be found they all work just fine, shouldn't they not be working if they really were dynamically linked to it?

**Reply by Biswa96** [reply]
You can use ntldd tool to find if the executable is really linked with stdc++. Install the [ntldd](https://packages.msys2.org/base/mingw-w64-ntldd-git) package in mingw. Or there is a GUI tool also, called [Dependencies](https://github.com/lucasg/dependencies).

**Reply by mati865** [reply]
Function sections are basically broken with GCC when targeting Windows.
They work fine in Clang. However just like on Linux to use them the libraries have to be built with them.

**Reply by mati865** [reply]
> Strangely when I try running the compiled binaries with libstdc++.dll in a directory where it can't be found they all work just fine, shouldn't they not be working if they really were dynamically linked to it?

Maybe there is `libstdc++.dll` in your `PATH`?
Just like Biswa96 said ntldd is the great tool to find dependencies.

**Comment by revelator** [comment]
aye i usually add -fno-function-sections -fno-data-sections when building gcc as these seem to be elf related.
